### PR TITLE
Add Sve.IsSupported support

### PR DESCRIPTION
**Pull request by a74nh**
Add the SVE API, but it only includes the IsSupported method.

Add the blank test files to go along with it

### CODE REVIEW

**Comment by a74nh** [comment]
This is cut from https://github.com/dotnet/runtime/pull/97695

I'm still a little confused about the testing. I can get the `.cs` files to generate. But then I'm not sure how to build and run the generated test files.

Also, what do I need to build+run to get `PrintSupportedIsa()` (in Shared/Program.cs) to print?

For now, the new SVE block inside `GenerateHWIntrinsicTests_Arm.cs` just contains a duplication of the existing `LoadUnOpTest.template`, which can be removed as soon as the first Sve API entry is added.

@kunalspathak @dotnet/arm64-contrib @tannergooding

**Comment by tannergooding** [comment]
The general process for tests is going to be:
1. Add the new table entries to https://github.com/dotnet/runtime/blob/main/src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs
2. Add an `Sve` folder to https://github.com/dotnet/runtime/tree/main/src/tests/JIT/HardwareIntrinsics/Arm which follows the general setup of https://github.com/dotnet/runtime/tree/main/src/tests/JIT/HardwareIntrinsics/Arm/AdvSimd (`Sve_r.csproj`, `Sve_ro.csproj`, `Program.Sve.cs`)
3. Build and run the tests according to https://github.com/dotnet/runtime/blob/main/docs/workflow/testing/coreclr/testing.md

File: src/coreclr/jit/hwintrinsiclistarm64sve.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+/*****************************************************************************/
+#ifndef HARDWARE_INTRINSIC
+#error Define HARDWARE_INTRINSIC before including this file
+#endif
+/*****************************************************************************/
+
+// clang-format off
+
+#ifdef FEATURE_HW_INTRINSICS
+// ***************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************
```
Comment (kunalspathak) [inline src/coreclr/jit/hwintrinsiclistarm64sve.h:13]: I thought `IsSupported` would have entry here, but yes, it's not needed. So, wondering if we should still get an empty file or have it as part of a PR that adds anything relevant in it?

**Review by kunalspathak — COMMENTED** [review]
@tannergooding - are there other places where we should set the bits for `IsSupported` or this should be good enough?

File: src/coreclr/jit/hwintrinsiclistarm64sve.h (hunk 1 above)
Comment (a74nh) [reply src/coreclr/jit/hwintrinsiclistarm64sve.h:13]: The only reason for keeping this here is it gets it done now - I can then write API PRs without being concerned about the dependencies of which one adds the files.

That includes the `hwintrinsiclistarm64.h` and `CMakeLists.txt` changes too.

File: src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs [hunk 2]
```diff
@@ -2920,6 +2928,9 @@ void ProcessInputs(string groupName, (string templateFileName, Dictionary<string
 
     Directory.CreateDirectory(outputDirectory);
 
+    Console.WriteLine($"{outputDirectory}");
```
[the remark this answers is not in the document]
Comment (a74nh) [reply src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs]: Oops. Yes, I thought I had removed these. Done

**Comment by a74nh** [comment]
Test results on SVE enabled machine running HardwareIntrinsics_Arm_r.sh:

```
14:29:58.994 Running test: JIT/HardwareIntrinsics/Arm/Sve/Sve_r/Sve_r.dll
Supported ISAs:
  AdvSimd:   True
  Aes:       True
  ArmBase:   True
  Crc32:     True
  Dp:        True
  Rdm:       True
  Sha1:      True
  Sha256:    True
  Sve:       True

Beginning scenario: RunBasicScenario_Load
Beginning scenario: RunReflectionScenario_Load
14:29:59.005 Passed test: JIT/HardwareIntrinsics/Arm/Sve/Sve_r/Sve_r.dll
```

**Comment by a74nh** [comment]
>2024-02-02T12:15:09.0048116Z ##[error].packages/microsoft.dotnet.apicompat.task/9.0.100-alpha.1.24072.3/build/Microsoft.DotNet.ApiCompat.ValidateAssemblies.Common.targets(16,5): error : (NETCORE_ENGINEERING_TELEMETRY=Build) API breaking changes found. If those are intentional, the APICompat suppression file can be updated by rebuilding with '/p:ApiCompatGenerateSuppressionFile=true'

I'm not sure how to fix this in the PR.

**Comment by kunalspathak** [comment]
> I'm not sure how to fix this in the PR.

there were few missing places. my changes should fix it.

File: src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs [hunk 3]
```diff
@@ -2893,6 +2893,13 @@
     ("SecureHashTernOpTest.template",     new Dictionary<string, string> { ["TestName"] = "ScheduleUpdate1_Vector128_UInt32",                                                                      ["Isa"] = "Sha256",        ["LoadIsa"] = "AdvSimd", ["Method"] = "ScheduleUpdate1",                                                      ["RetVectorType"] = "Vector128", ["RetBaseType"] = "UInt32",  ["Op1VectorType"] = "Vector128", ["Op1BaseType"] = "UInt32", ["Op2VectorType"] = "Vector128", ["Op2BaseType"] = "UInt32", ["Op3VectorType"] = "Vector128", ["Op3BaseType"] = "UInt32", ["LargestVectorSize"] = "16", ["NextValueOp1"] = "0x00112233",                               ["NextValueOp2"] = "0x44556677",                               ["NextValueOp3"] = "0x8899AABB",                               ["ExpectedResult"] = "{0x248F1BDF, 0x248F1BDF, 0xB303DDBA, 0xF74821FE}"}),
 };
 
+(string templateFileName, Dictionary<string, string> templateData)[] SveInputs = new []
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs:2896]: We already have just one file for x86 and this file is rarely touched/observed, with the different ISA collections being trivially collapsable, so I don't see a strong need.

But if others think it makes it easier to read/understand, by all means feel free.

File: src/coreclr/jit/hwintrinsiclistarm64sve.h (hunk 1 above)
Comment (tannergooding) [reply src/coreclr/jit/hwintrinsiclistarm64sve.h:13]: I think its better to get it done "now" as it will help avoid conflicts after

**Comment by kunalspathak** [comment]
> Test results on SVE enabled machine running HardwareIntrinsics_Arm_r.sh:
> 
> ```
> 14:29:58.994 Running test: JIT/HardwareIntrinsics/Arm/Sve/Sve_r/Sve_r.dll
> Supported ISAs:
>   AdvSimd:   True
>   Aes:       True
>   ArmBase:   True
>   Crc32:     True
>   Dp:        True
>   Rdm:       True
>   Sha1:      True
>   Sha256:    True
>   Sve:       True
> 
> Beginning scenario: RunBasicScenario_Load
> Beginning scenario: RunReflectionScenario_Load
> 14:29:59.005 Passed test: JIT/HardwareIntrinsics/Arm/Sve/Sve_r/Sve_r.dll
> ```

@fanyang-mono - how can we verify if this returns `false` for Mono?

**Comment by fanyang-mono** [comment]
@kunalspathak The change I suggested [here](https://github.com/dotnet/runtime/pull/97695#discussion_r1474711081) is needed, in order for Mono to return `false` for `IsSupported` check.

**Comment by vargaz** [comment]
Mono changes look ok to me.

**Comment by a74nh** [comment]
With the mono changes added I built mono and ran the testsuite according to mono/testing.md

Looking in the test logs:
```
13:13:26.469 Running test: JIT/HardwareIntrinsics/Arm/Sve/Sve_r/Sve_r.dll
Supported ISAs:
  AdvSimd:   False
  Aes:       False
  ArmBase:   True
  Crc32:     False
  Dp:        False
  Rdm:       False
  Sha1:      False
  Sha256:    False
  Sve:       False
```

**Review by kunalspathak — APPROVED** [review]
LGTM. Thanks!

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Sve.PlatformNotSupported.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+        [Intrinsic]
+        public new abstract class Arm64 : AdvSimd.Arm64
+        {
+            internal Arm64() { }
+
+            public static new bool IsSupported { get => IsSupported; }
+        }
+
+    }
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Sve.PlatformNotSupported.cs:31]: This needs to follow the existing pattern, including correctly returning constant false and only marking `IsSupported` as intrinsic:
```suggestion
    [CLSCompliant(false)]
    [System.Runtime.Versioning.RequiresPreviewFeaturesAttribute("Sve is in preview.")]
#if SYSTEM_PRIVATE_CORELIB
    public
#else
    internal
#endif
    abstract class Sve : AdvSimd
    {
        internal Sve() { }

        public static new bool IsSupported { [Intrinsic] get { return false; } }

        public new abstract class Arm64 : AdvSimd.Arm64
        {
            internal Arm64() { }

            public static new bool IsSupported { [Intrinsic] get { return false; } }
        }
    }
```

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Sve.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        public static new bool IsSupported { get => IsSupported; }
+
+        [Intrinsic]
+        public new abstract class Arm64 : AdvSimd.Arm64
+        {
+            internal Arm64() { }
+
+            public static new bool IsSupported { get => IsSupported; }
+        }
+
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Sve.cs:30]: nit: stray newline
```suggestion
```

File: src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs [hunk 6]
```diff
@@ -2893,6 +2893,13 @@
...
+{
+    //TODO-SVE: Replace LoadUnOpTest with SVE tests
+    ("LoadUnOpTest.template",             new Dictionary<string, string> { ["TestName"] = "LoadVector64_Byte",                                                                                     ["Isa"] = "AdvSimd",                                ["Method"] = "LoadVector64",                                                         ["RetVectorType"] = "Vector64",  ["RetBaseType"] = "Byte",    ["Op1VectorType"] = "Vector64",  ["Op1BaseType"] = "Byte",                                                                                                                             ["LargestVectorSize"] = "8",  ["NextValueOp1"] = "TestLibrary.Generator.GetByte()",                                                                                                                                                        ["ValidateIterResult"] = "firstOp[i] != result[i]"}),
```
Comment (tannergooding) [inline src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs:2899]: This entry is "dead" and should be excluded. We should add entries to the list when new APIs are actually added.

**Review by tannergooding — CHANGES_REQUESTED** [review]
The changes in general LGTM, but the Sve.PlatformNotSupported.cs file in particular needs to follow the correct pattern.

File: src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs (hunk 6 above)
Comment (a74nh) [reply src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs:2899]: Removed. That was just there to allow the test to build, but I've now fixed by using `Array.empty()`

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Sve.PlatformNotSupported.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+        public static new bool IsSupported { [Intrinsic] get { return false; } }
+
+        public new abstract class Arm64 : AdvSimd.Arm64
+        {
+            internal Arm64() { }
+
+            public static new bool IsSupported { [Intrinsic] get { return false; } }
+        }
+    }
```
Comment (kunalspathak) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Sve.PlatformNotSupported.cs:34]: There is a missing `}` here.

Comment (a74nh) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Sve.PlatformNotSupported.cs:34]: Fixed

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Sve.PlatformNotSupported.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+using System.Runtime.CompilerServices;
+using System.Runtime.Intrinsics;
+using System.Numerics;
+
+namespace System.Runtime.Intrinsics.Arm
+{
+    /// <summary>
+    /// This class provides access to the ARM SVE hardware instructions via intrinsics
+    /// </summary>
+    [Intrinsic]
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Sve.PlatformNotSupported.cs:14]: The `PlatformNotSupported.cs` version shouldn't be marked `Intrinsic` here, it should only be marked on the `IsSupported` APIs.
```suggestion
-    [Intrinsic]
```

Comment (a74nh) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Sve.PlatformNotSupported.cs]: Fixed.

**Review by tannergooding — APPROVED** [review]

**Comment by kunalspathak** [comment]
All failures are known:

image

**Comment by kunalspathak** [comment]
> Test results on SVE enabled machine running HardwareIntrinsics_Arm_r.sh:
> 
> ```
> 14:29:58.994 Running test: JIT/HardwareIntrinsics/Arm/Sve/Sve_r/Sve_r.dll
> Supported ISAs:
>   AdvSimd:   True
>   Aes:       True
>   ArmBase:   True
>   Crc32:     True
>   Dp:        True
>   Rdm:       True
>   Sha1:      True
>   Sha256:    True
>   Sve:       True
> 
> Beginning scenario: RunBasicScenario_Load
> Beginning scenario: RunReflectionScenario_Load
> 14:29:59.005 Passed test: JIT/HardwareIntrinsics/Arm/Sve/Sve_r/Sve_r.dll
> ```

It just occurred to me that this should return `false` on windows/arm64 SVE machine until windows support is not added.

**Comment by tannergooding** [comment]
> It just occurred to me that this should return false on windows/arm64 SVE machine until windows support is not added.

Shouldn't that already fall out from the VM support?

**Comment by tannergooding** [comment]
We set it for Unix here: https://github.com/dotnet/runtime/blob/main/src/native/minipal/cpufeatures.c#L362-L363

and never set it for Windows, as no way to check the API exists yet:
https://github.com/dotnet/runtime/blob/main/src/native/minipal/cpufeatures.c#L406-L439

**Comment by kunalspathak** [comment]
> We set it for Unix here: https://github.com/dotnet/runtime/blob/main/src/native/minipal/cpufeatures.c#L362-L363
> 
> and never set it for Windows, as no way to check the API exists yet: https://github.com/dotnet/runtime/blob/main/src/native/minipal/cpufeatures.c#L406-L439

perfect. Thanks for confirming.

**Pull request merged**